model/host/db.go

```diff
@@ -314,7 +315,26 @@ func allHostsSpawnedByFinishedTasks() ([]Host, error) {
 			"as":           runningTasks,
 		}},
 		{"$unwind": "$" + runningTasks},
-		{"$match": bson.M{bsonutil.GetDottedKeyName(runningTasks, task.StatusKey): bson.M{"$in": evergreen.CompletedStatuses}}},
+		{"$match": bson.M{
+			"$or": []bson.M{
+				// If the task is finished, then its host should be terminated.
+				{
+					bsonutil.GetDottedKeyName(runningTasks, task.StatusKey): bson.M{"$in": evergreen.CompletedStatuses},
+				},
+
+				// If the task execution number is greater than the host's task
+				// execution number, than the host belongs to an old task and should
+				// be terminated.
+				{
```

Comment:
Wouldn't the host already be terminated at this point, since the previous task execution should be done by the time a new execution is started? 
In other words, shouldn't "their tasks have finished" be true every time "the task execution number is greater than the host's task execution number" is true?

Reply:
Evergreen has kind of an odd way of handling task restarts. When a task is restarted, it moved the task from the "tasks" to the "old_tasks" collection and creates a new task with an incremented execution number in the "tasks".

Separately, `PopulateHostTerminationJobs` runs once per minute. If a task finishes and a host termination job for the host created by the task runs before a person restarts the task, then yes, the host will be terminated. It is unlikely a person will restart a task by the time the host termination job runs, since they run every minute.

However, Evergreen itself can also restart a task. It restarts a task if the host that the task is running on is terminated by AWS. In that case, Evergreen marks the task finished and then immediately restarts it, which means that the time that the task is actually in the "success" state is very small, because it immediately gets put in the old_tasks collection and a new, undispatched task is created in the tasks colleciton.

I believe the reason that users are only starting to notice this now is because of the increased rate of host termination by AWS.